federation/pkg/federation-controller/util/test/test_helper.go

```diff
@@ -44,77 +44,67 @@ func (wd *WatcherDispatcher) register(watcher *watch.FakeWatcher) {
 	defer wd.Unlock()
 	wd.watchers = append(wd.watchers, watcher)
 	for _, event := range wd.eventsSoFar {
-		go watcher.Action(event.Type, event.Object)
+		watcher.Action(event.Type, event.Object)
 	}
 }
 
 // Add sends an add event.
 func (wd *WatcherDispatcher) Add(obj runtime.Object) {
 	wd.Lock()
 	defer wd.Unlock()
-	event := &watch.Event{
-		Type:   watch.Added,
-		Object: obj,
-	}
-	wd.eventsSoFar = append(wd.eventsSoFar, event)
+	wd.eventsSoFar = append(wd.eventsSoFar, &watch.Event{Type: watch.Added, Object: obj})
 	for _, watcher := range wd.watchers {
-		go watcher.Add(obj)
+		if !watcher.IsStopped() {
```

Comment:
There is still possible race here, because watcher can call Stop at any time. In particular it can call Stop between calling IsStopped and Add here.